File: FBDeviceControl/Commands/FBDeviceSocketForwardingCommands.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBDeviceSocketForwardingCommands.h"

#import "FBDevice.h"
#import "FBDeviceControlError.h"

@interface FBDeviceSocketForwardingCommands ()

@property (nonatomic, weak, readonly) FBDevice *device;

@end

@implementation FBDeviceSocketForwardingCommands

#pragma mark Initializers

+ (instancetype)commandsWithTarget:(FBDevice *)target
{
  return [[self alloc] initWithDevice:target];
}

- (instancetype)initWithDevice:(FBDevice *)device
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _device = device;

  return self;
}

#pragma mark FBSocketForwardingCommands Implementation

- (FBFuture<NSNull *> *)drainLocalFileInput:(int)localFileDescriptorInput localFileOutput:(int)localFileDescriptorOutput remotePort:(int)remotePort
{
  NSError *error = nil;
  id<FBDataConsumer> localConsumer = [FBFileWriter asyncWriterWithFileDescriptor:localFileDescriptorOutput closeOnEndOfFile:NO error:&error];
  if (!localConsumer) {
    return [FBFuture futureWithError:error];
  }
  return [[self
    consumerForRemotePort:remotePort writingTo:localConsumer]
    onQueue:self.device.asyncQueue pop:^(id<FBDataConsumer> remoteConsumer) {
      id<FBFileReader> reader = [FBFileReader readerWithFileDescriptor:localFileDescriptorInput closeOnEndOfFile:NO consumer:remoteConsumer logger:nil];
      return [[reader
        startReading]
        onQueue:self.device.asyncQueue fmap:^(id _) {
          return reader.finishedReading;
        }];
    }];
}

#pragma mark Private

- (FBFutureContext<id<FBDataConsumer>> *)consumerForRemotePort:(int)remotePort writingTo:(id<FBDataConsumer>)consumer
{
  return [[self
    localSocketFromRemotePort:remotePort]
    onQueue:self.device.asyncQueue pend:^(NSNumber *remoteSocket) {
      NSError *error = nil;
      id<FBDataConsumer> writer = [FBFileWriter asyncWriterWithFileDescriptor:remoteSocket.intValue closeOnEndOfFile:NO error:&error];
      if (!writer) {
        return [FBFuture futureWithError:error];
      }
      id<FBFileReader> reader = [FBFileReader readerWithFileDescriptor:remoteSocket.intValue closeOnEndOfFile:NO consumer:consumer logger:nil];
      return [[reader
        startReading]
        mapReplace:writer];
    }];
}

- (FBFutureContext<NSNumber *> *)localSocketFromRemotePort:(int)remotePort
{
  id<FBControlCoreLogger> logger = self.device.logger;
  return [[[self.device
    connectToDeviceWithPurpose:@"Socket Connection"]
    onQueue:self.device.workQueue pop:^(id<FBDeviceCommands> device) {
      int connectionID = device.calls.GetConnectionID(device.amDeviceRef);
      if (connectionID <= 0) {
        return [[FBDeviceControlError
          describeFormat:@"Failed to get ConnectionID from Device"]
          failFuture];
      }
      [logger logFormat:@"Got connection ID %d, for device. Connecting to remote port %d", connectionID, remotePort];
      int localSocket = 0;
      int status = device.calls.USBMuxConnectByPort(connectionID, htons(remotePort), &localSocket);
      if (status != 0) {
        return [[FBDeviceControlError
          describeFormat:@"Failed to connect to remote port %d", remotePort]
          failFuture];
      }
      [logger logFormat:@"Got local socket %d for remote port %d", localSocket, remotePort];
      return [FBFuture futureWithResult:@(localSocket)];
    }]
    onQueue:self.device.asyncQueue contextualTeardown:^(NSNumber *localSocketNumber, FBFutureState _) {
      [logger logFormat:@"Closing local socket %@", localSocketNumber];
      close(localSocketNumber.intValue);
      return FBFuture.empty;
    }];
}

@end

File: FBDeviceControl/Commands/FBDeviceSocketForwardingCommands.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@preconcurrency import FBControlCore
import Foundation

// swiftlint:disable force_unwrapping

@objc(FBDeviceSocketForwardingCommands)
public class FBDeviceSocketForwardingCommands: NSObject, FBiOSTargetCommand {
  private(set) weak var device: FBDevice?

  // MARK: Initializers

  @objc public class func commands(with target: any FBiOSTarget) -> Self {
    return unsafeDowncast(FBDeviceSocketForwardingCommands(device: target as! FBDevice), to: self)
  }

  init(device: FBDevice) {
    self.device = device
    super.init()
  }

  // MARK: FBSocketForwardingCommands (legacy FBFuture entry point)

  public func drainLocalFileInput(
    _ localFileDescriptorInput: Int32,
    localFileOutput localFileDescriptorOutput: Int32,
    remotePort: Int32
  ) -> FBFuture<NSNull> {
    fbFutureFromAsync { [self] in
      try await drainLocalFileInputAsync(
        localFileDescriptorInput,
        localFileOutput: localFileDescriptorOutput,
        remotePort: remotePort)
      return NSNull()
    }
  }

  // MARK: - Async

  fileprivate func drainLocalFileInputAsync(
    _ localFileDescriptorInput: Int32,
    localFileOutput localFileDescriptorOutput: Int32,
    remotePort: Int32
  ) async throws {
    guard let device else {
      throw FBDeviceControlError().describe("Device is nil").build()
    }
    var error: NSError?
    guard let localConsumer = FBFileWriter.asyncWriter(withFileDescriptor: localFileDescriptorOutput, closeOnEndOfFile: false, error: &error) else {
      throw error!
    }
    try await withFBFutureContext(device.connectToDevice(withPurpose: "Socket Connection")) { connectedDevice in
      let localSocket = try Self.openLocalSocket(toRemotePort: Int(remotePort), on: connectedDevice, logger: device.logger)
      defer {
        device.logger?.log("Closing local socket \(localSocket)")
        close(localSocket)
      }
      var writerError: NSError?
      guard let remoteWriter = FBFileWriter.asyncWriter(withFileDescriptor: localSocket, closeOnEndOfFile: false, error: &writerError) else {
        throw writerError!
      }
      let remoteReader = FBFileReader.reader(withFileDescriptor: localSocket, closeOnEndOfFile: false, consumer: localConsumer, logger: nil)
      try await bridgeFBFutureVoid(remoteReader.startReading())

      let inputReader = FBFileReader.reader(withFileDescriptor: localFileDescriptorInput, closeOnEndOfFile: false, consumer: remoteWriter, logger: nil)
      try await bridgeFBFutureVoid(inputReader.startReading())
      _ = try await bridgeFBFuture(inputReader.finishedReading)
    }
  }

  // MARK: Private

  private static func openLocalSocket(toRemotePort remotePort: Int, on device: any FBDeviceCommands, logger: (any FBControlCoreLogger)?) throws -> Int32 {
    guard let getConnectionID = device.calls.GetConnectionID else {
      throw FBDeviceControlError.describe("GetConnectionID not available").build()
    }
    let connectionID = getConnectionID(device.amDeviceRef)
    if connectionID <= 0 {
      throw FBDeviceControlError.describe("Failed to get ConnectionID from Device").build()
    }
    logger?.log("Got connection ID \(connectionID), for device. Connecting to remote port \(remotePort)")
    var localSocket: Int32 = 0
    guard let usbMuxConnect = device.calls.USBMuxConnectByPort else {
      throw FBDeviceControlError.describe("USBMuxConnectByPort not available").build()
    }
    let status = usbMuxConnect(connectionID, Int32(UInt16(remotePort).bigEndian), &localSocket)
    if status != 0 {
      throw FBDeviceControlError.describe("Failed to connect to remote port \(remotePort)").build()
    }
    logger?.log("Got local socket \(localSocket) for remote port \(remotePort)")
    return localSocket
  }
}
